components/script/dom/bindings/str.rs

```diff
@@ -208,6 +208,77 @@ impl DOMString {
         self.0.truncate(last_non_whitespace);
         let _ = self.0.splice(0..first_non_whitespace, "");
     }
+
+    /// Validates this `DOMString` is a time string according to
+    /// <https://html.spec.whatwg.org/multipage/#valid-time-string>.
+    pub fn is_valid_time_string(&self) -> bool {
+        enum State {
+            HourHigh,
+            HourLow09,
+            HourLow03,
+            MinuteColon,
+            MinuteHigh,
+            MinuteLow,
+            SecondColon,
+            SecondHigh,
+            SecondLow,
+            MilliStop,
+            MilliHigh,
+            MilliMiddle,
+            MilliLow,
+            Done,
+            Error,
+        }
+        let next_state = |valid: bool, next: State| -> State { if valid { next } else { State::Error } };
+
+        let state = self.chars().fold(State::HourHigh, |state, c| {
```

Comment:
One thing that we didn't check here is whether the character is ASCII. If at any point we run into a non-ASCII character, we would want to error out. In addition, I recall that iterating over chars is actually a bit slow, due to the fact that it needs to ensure UTF-8 well-formedness on each iteration.

So instead, maybe we could iterate over the underlying `u8` data type by calling `as_bytes`, and importing `AsciiExt` so that we can use the `.is_ascii()` function to check whether the character is ASCII on each iteration.

Reply:
I don't know if Unicode has other digits that are valid for `is_digit`, if not we are already check for ASCII implicit.
When using bytes (`u8`) there should be no need to check for ASCII as we already check for specific values (`0-9`, `:`, `.`) which implies there are only ASCII values.

Reply:
Good point. I'll leave it up to you whether you want to use `u8` or `char` in this PR.